From 5d81330be0c6b69d2c8a4adfae8916f043b9a1f4 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Tue, 22 Jul 2025 13:07:25 +0200
Subject: [PATCH] net: phy: smsc: handle WoL event upon receiving an interrupt

If an interrupt caused by a WoL event is received whilst not being
in a low-power state, the interrupt handler must clear the WoL flags
to be able to handle other WoL events.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I75246f4987e8cdfd0894ead2966d3b32477a5217
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/476533
Reviewed-by: Christophe ROULLIER <christophe.roullier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Christophe ROULLIER <christophe.roullier@foss.st.com>
Domain-Review: Christophe ROULLIER <christophe.roullier@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 drivers/net/phy/smsc.c  | 22 ++++++++++++++++++++++
 include/linux/smscphy.h |  2 ++
 2 files changed, 24 insertions(+)

diff --git a/drivers/net/phy/smsc.c b/drivers/net/phy/smsc.c
index 454c3df5371f..323a8e952914 100644
--- a/drivers/net/phy/smsc.c
+++ b/drivers/net/phy/smsc.c
@@ -110,6 +110,28 @@ irqreturn_t smsc_phy_handle_interrupt(struct phy_device *phydev)
 		return IRQ_NONE;
 	}
 
+	/* Handle WoL events */
+	if (irq_status & MII_LAN874X_ISF_INT8) {
+		int rc = phy_read_mmd(phydev, MDIO_MMD_PCS, MII_LAN874X_PHY_MMD_WOL_WUCSR);
+
+		if (rc < 0) {
+			phy_error(phydev);
+			return IRQ_NONE;
+		}
+
+		if (!(rc & MII_LAN874X_PHY_WOL_STATUS_MASK)) {
+			phy_error(phydev);
+			return IRQ_NONE;
+		}
+
+		rc = phy_write_mmd(phydev, MDIO_MMD_PCS, MII_LAN874X_PHY_MMD_WOL_WUCSR,
+				   rc | MII_LAN874X_PHY_WOL_STATUS_MASK);
+		if (rc < 0) {
+			phy_error(phydev);
+			return IRQ_NONE;
+		}
+	}
+
 	if (!(irq_status & MII_LAN83C185_ISF_INT_PHYLIB_EVENTS))
 		return IRQ_NONE;
 
diff --git a/include/linux/smscphy.h b/include/linux/smscphy.h
index cdf266960032..7e00df78bf04 100644
--- a/include/linux/smscphy.h
+++ b/include/linux/smscphy.h
@@ -74,4 +74,6 @@ int smsc_phy_probe(struct phy_device *phydev);
 
 #define MII_LAN874X_PHY_PME_SELF_CLEAR_DELAY	0x1000 /* 81 milliseconds */
 
+#define MII_LAN874X_ISF_INT8 (1<<8)		/* WoL event */
+
 #endif /* __LINUX_SMSCPHY_H__ */
-- 
2.53.0

